infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java

```diff
@@ -89,92 +111,74 @@ public PollReport poll(
                 layout.getEpoch(), layout.getClusterId(), allServers, routers, sequencerMetrics,
                 ImmutableList.copyOf(layout.getUnresponsiveServers()),
-                fileSystemStats
-        );
+                fileSystemStats);
     }
 
     private Map<String, IClientRouter> adjustRouters(CorfuRuntime corfuRuntime, Set<String> allServers) {
         // Set up arrays for routers to the endpoints.
         Map<String, IClientRouter> routers = new HashMap<>();
         allServers.forEach(server -> {
             IClientRouter router = corfuRuntime.getRouter(server);
-            router.setTimeoutResponse(networkStretcher.getCurrentPeriod().toMillis());
             routers.put(server, router);
         });
         return routers;
     }
 
-    /**
-     * PollRound consists of iterations. In each iteration, the FailureDetector pings all the
-     * responsive nodes in the layout and also collects their {@link NodeState}-s to provide {@link ClusterState}.
-     * Failure detector collect a number of {@link PollReport}-s equals to failureThreshold number then
-     * aggregates final {@link PollReport}.
-     * Aggregation step:
-     * - go through all reports
-     * - aggregate all wrong epochs from all intermediate poll reports then remove all responsive nodes from the list.
-     * If wrongEpochsAggregated variable still contains wrong epochs
-     * then it will be corrected by {@link RemoteMonitoringService}
-     * - Aggregate connected and failed nodes from all reports
-     * - Timeouts tuning
-     * - provide a poll report based on latest pollIteration report and aggregated values
-     *
-     * @return Poll Report with detected failed nodes and out of phase epoch nodes.
-     */
-    @VisibleForTesting
-    PollReport pollRound(
-            long epoch, UUID clusterID, Set<String> allServers, Map<String, IClientRouter> router,
-            SequencerMetrics sequencerMetrics, ImmutableList<String> layoutUnresponsiveNodes,
-            FileSystemStats fileSystemStats) {
-
-        if (failureThreshold < 1) {
-            throw new IllegalStateException("Invalid failure threshold");
-        }
-
+    PollReport pollRoundExpBackoff(long epoch, UUID clusterID, Set<String> allServers, Map<String, IClientRouter> router,
+                                   SequencerMetrics sequencerMetrics, ImmutableList<String> layoutUnresponsiveNodes,
+                                   FileSystemStats fileSystemStats) {
+
+        final PollConfig pollConfig = this.pollConfig.orElseGet(() -> PollConfig.builder().build());
+        Consumer<MixedBoundRetry> retrySettings = settings -> {
+            settings.setOverallMaxRetryDuration(pollConfig.getMaxDetectionDuration());
+            settings.setMaxRetryDuration(pollConfig.getMaxSleepBetweenRetries());
+            settings.setBaseDuration(pollConfig.getInitSleepBetweenRetries());
+            settings.setRandomPart(pollConfig.getJitterFactor());
+        };
```

Comment:
For fixed interval heart-beating, you poll at a fixed interval and expect at least 1 response within some time window. When would this strategy be better ?

Exponential back off is usually used to randomize an unknown number of participants to increase the likelihood of a response, but in the case, no response/timeout is also a valid signal for the fault detector. 

Essentially, it's not clear to me why using exponential back off is a better approximation of "health" than fixed interval.

Reply:
My view was that the exponential backoff provides a better balance
between faster detection times and stability. Setting the fixed interval for some environments
would result in potentially slower detection rates (where finding a failure could be done quicker) and for others might cause stability issues (where the poll on the other hand needs to be longer). This exponential backoff covers this spectrum by configuring the backoff parameters but it also sets the cap on the overall duration of the poll round. So if there are fewer errors in the systems, they are detected fast enough, if there are more failures in the system, it increases the poll round to slow down the detection rates to aid stability.

Reply:
I still don't see how it can be faster. Can you provide a polling schedule that shows you can detect a failed node faster than fixed polling when both are bounded by the same max timeout window.

For example, a polling scheduling can look something like this (max timeout 5 seconds and 2 seconds polling):

Max duration:  |--------5---------| 
Fixed polling:   |---2---|---2---|---2---| (3 rounds)
Expo polling:   |---2---|-------4-------| (2 rounds)

Wouldn't both of these methods detect the failure within 5 seconds, or am I missing something?

Reply:
This failure detector algorithm runs poll rounds for some amount of time and then after poll rounds are done, it runs a failure detection/healing algorithm.

Say we have three cases:
Case A (System is mostly stable, we know that at most one failure among all the poll rounds is detected). 
Case B (System is mostly unstable, we know that we detect a failure on every poll round most of the time, say, because of some persistent infra issue). 
Case C (System is mixed, we don't know about the failure distribution. Sometimes we detect at most one failure among all the poll rounds, sometimes we detect a failure on every poll round, sometimes we get something in between).

Case A:
Since we know the system is mostly stable, we want to detect the failures as quickly as possible, because we know they are real actionable failures. We set the max timeout to 5 seconds. Both approaches will detect the failure within the 5-second window. Cluster reconfigurations are desirable and will happen infrequently and fast. 

Case B: 
Since we know the system is mostly unstable, we want to detect failures much less frequently, giving some buffer time of stability before the FD algorithm kicks in.  We set the max timeout to 30 seconds. Both approaches will be detecting a failure after 30 seconds. Cluster reconfigurations are undesirable but at least will be happening every 30 seconds. If we set the timeout to 5 seconds here, we will be detecting "failures" all the time, reconfiguring cluster all the time, joining and kicking out nodes very frequently when this can be avoided/done less frequently. 

Case C:
Say as a compromise we set the max timeout to 15 seconds for both.  If system C is more like system A, interval retry will spend 15 seconds detecting a failure while mixed retry will spend only (1 + 2 + 2) = 5 seconds (since only one failure happened). If system C is more like system B, both will give the needed buffer of 15 seconds for the system to stabilize. If it's somewhere in between system A and system B, mixed retry will always do better than 15 seconds.

Reply:
There seems to be two issues: there is an assumption that the "stability" condition can be inferred from `currReport.getFailedNodes().isEmpty()` and that requests for failed nodes will fail-fast, both of those assumptions are incorrect.

For the first case, imagine that 1 node out of 3 will always be offline, this means that detection time for the other two will be skewed.

For the second case, imagine that the request to the "unresponsive node" doesn't fail fast and needs 5 seconds to timeout. That means each time `pollAsync` is called, 5 seconds (i.e., request timeout) will elapse because of `CFUtils.allOf(clusterState.values()).join();`, in that case it will take:

request timeout + nextWait + request timeout + nextWait + request timeout  = 5 + 1 + 5 + 1+ 5 = 17 seconds 

If there exists a node that has been detected from a previous run as unresponsive then the detection time will be dilated further.

Reply:
@Maithem 
Regarding the first case, how can an FD distinguish between the node that is always down and the node that has failed by sending the requests?  
Regarding the second case, in my testing, I could never hit the request timeout, only the connection timeout; all the requests failed fast. Do you have a test/repro that reproduces the issue in which you can connect but then hit the request timeout?

Reply:
> Regarding the first case, how can an FD distinguish between the node that is always down and the node that has failed by sending the requests?

That's the point, you can't and I think there shouldn't be a special cases.

> Regarding the second case, in my testing, I could never hit the request timeout, only the connection timeout; all the requests failed fast. Do you have a test/repro that reproduces the issue in which you can connect but then hit the request timeout?

This can happen in various ways, there could be asymmetric or partial network partitions, or that remote peers crash  without cleanly closing the connection causing a half-open tcp connection [1], excessive GC on the server, etc...

You can simulate an rpc request timeout by adding an iptables that drops the response


[1] https://en.wikipedia.org/wiki/TCP_half-open

Reply:
> in my testing, I could never hit the request timeout

This is probably due to the fact that killing the process isn't sufficient to cause a half-open tcp connection because the kernel will terminate the connection properly during cleanup.

Reply:
@Maithem 
Also, for the first case, what do you suggest? The assumption `currReport.getFailedNodes().isEmpty()`  is there in the first place to artificially increase the timeouts when the nodes were sampled to appear mostly unresponsive during polling. Because if we leave the timeout short and we do not make the assumption about stability, and we can not say if it's a dead node or just not responding node, we end up in a situation where the cluster becomes flaky because we get more requests that go through when the very flaky node tries to heal. I believe the patch was introduced because @medhavidhawan asked @xnull to increase the timeouts to reduce this flakiness as it impacted the clients. How do you suggest leaving detection timeouts short but coming up with a strategy to consider specific nodes more "flaky" than others? Or are you saying it's fine, given that we have delta-reset? We have not yet tested the impact of delta-reset in production.

Regarding the second case, the poll round will not be scheduled since the beginning of the run if the run takes more than MAX_DETECTION_DURATION even with the 5-second timeouts present.  In that case, we will proceed to update the cluster view with any number of the accumulated reports. Yes, there can be a case in which, let's say, MAX_DETECTION_DURATION is 15 seconds, and at second 14, we encounter the RPC timeout that adds 5-second delay, which will yield a 19-second delay instead of 14, but the next round will not be scheduled, which I think is fine.

Reply:
The point of this change is to address the following two issues:
1. Make the parameters have a deterministic outcome
2. Use (1) to shorten the length of failure detection and reconfiguration

If this PR doesn't achieve both of those goals then what is the point of it? The non-determinism comes from the variable timeouts, this PR doesn't remove that, it just pushes it from the request level to the sleep between the polls.

The simplest form of FD is to use fixed heart beats, for example, for a timeout period of 6 seconds and a poll period of 2 seconds you have the following schedule:

poll----2----poll----2----poll----2----timeout (i.e., make decision)

That way the timeout period reflects execrably the time a decision will be made, that is, more deterministic than what's there and what this PR implements

Reply:
@Maithem 
You still need to answer my questions. 
```
How do you suggest leaving detection timeouts short but coming up with a strategy to consider specific nodes more "flaky" than others? 
```
Making the timeout short (6 seconds) will result in continuous cluster reconfigurations and instability in many environments where we currently run the DB. NetworkStretcher was designed to prevent that initially. This PR addresses the issues with NetworkStretcher, making it more deterministic and bound and removing the unnecessary RPC timeout tuning. 

Also, the max 30-second (double poll) heal timeout is reduced to the max of 15 seconds.

Reply:
> How do you suggest leaving detection timeouts short but coming up with a strategy to consider specific nodes more "flaky" than others? 

It is obvious that there is a trade-off between the timeout period and making false positive decisions. I'm not suggesting to shorten the timeout period while maintaining accuracy. My broader point is don't optimize for a pathological case and use the simplest form of polling to improve determinism. 

If there is a "flaky" node or a link, then we should address that with a different mechanism. For example, enforce a cool down period before the node rejoins. These techniques can only help you so much, but there are pathological cases where we shouldn't optimize for because they make the average case worse.

Also, since you haven't considered the request timeout contribution to the overall timeout the max timeout you mention in multiple responses is incorrect, which strengthens the point i'm trying to make having deterministic improves the reasoning model when we debug these issues.

Reply:
Ok, fair enough. I will replace this mechanism with a fixed poll and shorten the timeout period. With delta transfer merged, these reconfigurations should be alright anyway in the worst-case scenario.

Reply:
I think fixed polling is a reasonable choice, but if you want to explore this further I suggest you look at known algorithms: https://cwiki.apache.org/confluence/display/ZOOKEEPER/GSoCFailureDetector  , the original idea of a "network stretcher" is really just a botched attempt to do adaptive FD.